🐞 Engine panic when trying to access private registry

### What is the issue?

I am trying to run a container from a private registry (auth'd through `docker login` before hand), and the engine is dying with a panic


### Dagger version

dagger 0.19.0

### Steps to reproduce

```dagger core container from --address=$PRIVATE_REGISTRY/fastly/base-noble:latest terminal --cmd bash```

-- Engine crashes (see logs output)

```docker pull $PRIVATE_REGISTRY_URL/fastly/base-noble:latest``` works fine

### Log output

```
time="2025-10-07T18:01:04Z" level=debug msg="handling http request" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p contentType= method=GET path=/sessionAttachables session_id=shtidr95l79rth97nkewbp5hl span=e0e814c506b1756e spanID=e0e814c506b1756e trace=9843d8e30770f2222251382c179e4546 traceID=9843d8e30770f2222251382c179e4546 upgradeHeader=
time="2025-10-07T18:01:04Z" level=debug msg="buildkit session.Manager handleConn" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p session.id=b01jcreerrmg22tx8195l7ym1 session_id=shtidr95l79rth97nkewbp5hl spanID=e0e814c506b1756e traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:04Z" level=debug msg="buildkit session.Manager handleConn connected successfully" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p session.id=b01jcreerrmg22tx8195l7ym1 session_id=shtidr95l79rth97nkewbp5hl spanID=e0e814c506b1756e traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:04Z" level=debug msg="session manager handling conn ywdke3is250w9r67n586euj6p" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p session_id=shtidr95l79rth97nkewbp5hl spanID=e0e814c506b1756e traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:04Z" level=debug msg="buildkit session.Manager handleConn" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p session.id=ywdke3is250w9r67n586euj6p session_id=shtidr95l79rth97nkewbp5hl spanID=e0e814c506b1756e traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:04Z" level=debug msg="buildkit session.Manager handleConn connected successfully" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p session.id=ywdke3is250w9r67n586euj6p session_id=shtidr95l79rth97nkewbp5hl spanID=e0e814c506b1756e traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:04Z" level=debug msg="handling http request" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p contentType= method=GET path=/v1/traces session_id=shtidr95l79rth97nkewbp5hl span=3f453e62ed853a57 spanID=3f453e62ed853a57 trace=9843d8e30770f2222251382c179e4546 traceID=9843d8e30770f2222251382c179e4546 upgradeHeader=
time="2025-10-07T18:01:04Z" level=debug msg="handling http request" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p contentType= method=GET path=/v1/logs session_id=shtidr95l79rth97nkewbp5hl span=80e10bed05a4313b spanID=80e10bed05a4313b trace=9843d8e30770f2222251382c179e4546 traceID=9843d8e30770f2222251382c179e4546 upgradeHeader=
time="2025-10-07T18:01:04Z" level=debug msg="handling http request" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p contentType= method=GET path=/v1/metrics session_id=shtidr95l79rth97nkewbp5hl span=9f2167eb462cfcaa spanID=9f2167eb462cfcaa trace=9843d8e30770f2222251382c179e4546 traceID=9843d8e30770f2222251382c179e4546 upgradeHeader=
time="2025-10-07T18:01:04Z" level=debug msg="handling http request" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p contentType=application/json method=POST path=/query session_id=shtidr95l79rth97nkewbp5hl span=96bd88ecd9be9701 spanID=96bd88ecd9be9701 trace=9843d8e30770f2222251382c179e4546 traceID=9843d8e30770f2222251382c179e4546 upgradeHeader=
time="2025-10-07T18:01:04Z" level=debug msg="load cache for mkfile /schema.json with 5v1iqj0yxk968eivvsbc967dp::zgsewh2i8iib70yl4ymcdu8lc"
time="2025-10-07T18:01:04Z" level=debug msg="GetContentHashKey reusing ref zgsewh2i8iib70yl4ymcdu8lc with digest sha256:09f9fae15694fb63f79614bc565d1c71de4c5ce0e1bc3ae271fa88f329644921" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p session_id=shtidr95l79rth97nkewbp5hl spanID=57f38f7a4c24c616 traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:04Z" level=debug msg="handling http request" client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p contentType=application/json method=POST path=/query session_id=shtidr95l79rth97nkewbp5hl span=59ea80d02366de54 spanID=59ea80d02366de54 trace=9843d8e30770f2222251382c179e4546 traceID=9843d8e30770f2222251382c179e4546 upgradeHeader=
time="2025-10-07T18:01:04Z" level=debug msg="checked for cached auth handler namespace" cached=false key="PRIVATE_REGISTRY/fastly/base-focal::pull" name=PRIVATE_REGISTRY/fastly/base-focal scope=pull
time="2025-10-07T18:01:05Z" level=debug msg=Unauthorized client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p header="Bearer realm=\"https://PRIVATE_REGISTRY/service/token\",service=\"harbor-registry\",scope=\"repository:fastly/base-focal:pull\"" session_id=shtidr95l79rth97nkewbp5hl span="resolving PRIVATE_REGISTRY/fastly/base-focal:latest" spanID=25b0cb0a24b3fc8a traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:06Z" level=debug msg="checked for cached auth handler namespace" cached=true key="PRIVATE_REGISTRY/fastly/base-focal::pull" name=PRIVATE_REGISTRY/fastly/base-focal scope=pull
time="2025-10-07T18:01:06Z" level=debug msg="checked for cached auth handler namespace" cached=true key="PRIVATE_REGISTRY/fastly/base-focal::pull" name=PRIVATE_REGISTRY/fastly/base-focal scope=pull
time="2025-10-07T18:01:06Z" level=debug msg="checked for cached auth handler namespace" cached=true key="PRIVATE_REGISTRY/fastly/base-focal::pull" name=PRIVATE_REGISTRY/fastly/base-focal scope=pull
time="2025-10-07T18:01:06Z" level=debug msg="checked for cached auth handler namespace" cached=true key="PRIVATE_REGISTRY/fastly/base-focal::pull" name=PRIVATE_REGISTRY/fastly/base-focal scope=pull
time="2025-10-07T18:01:06Z" level=debug msg=Unauthorized client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p header="Bearer realm=\"https://PRIVATE_REGISTRY/service/token\",service=\"harbor-registry\",scope=\"repository:fastly/base-focal:pull\"" session_id=shtidr95l79rth97nkewbp5hl spanID=0e1477b1c966ce18 traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:07Z" level=debug msg=Unauthorized client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p header="Bearer realm=\"https://PRIVATE_REGISTRY/service/token\",service=\"harbor-registry\",scope=\"repository:fastly/base-focal:pull\"" session_id=shtidr95l79rth97nkewbp5hl spanID=0e1477b1c966ce18 traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:07Z" level=debug msg=Unauthorized client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p header="Bearer realm=\"https://PRIVATE_REGISTRY/service/token\",service=\"harbor-registry\",scope=\"repository:fastly/base-focal:pull\"" session_id=shtidr95l79rth97nkewbp5hl spanID=0e1477b1c966ce18 traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:07Z" level=debug msg=Unauthorized client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p header="Bearer realm=\"https://PRIVATE_REGISTRY/service/token\",service=\"harbor-registry\",scope=\"repository:fastly/base-focal:pull\"" session_id=shtidr95l79rth97nkewbp5hl spanID=0e1477b1c966ce18 traceID=9843d8e30770f2222251382c179e4546
time="2025-10-07T18:01:07Z" level=debug msg=Unauthorized client_hostname=G27H4H2CNV client_id=ywdke3is250w9r67n586euj6p header="Basic realm=\"harbor\"" session_id=shtidr95l79rth97nkewbp5hl spanID=0e1477b1c966ce18 traceID=9843d8e30770f2222251382c179e4546
dnsmasq[34]: read /etc/hosts - 18 names
dnsmasq[34]: read /var/run/containers/cni/dnsname/dagger/addnhosts - 2 names
dnsmasq[34]: read /etc/hosts - 18 names
dnsmasq[34]: read /var/run/containers/cni/dnsname/dagger/addnhosts - 1 names
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc7c5f4]

goroutine 106846 [running]:
github.com/dagger/dagger/engine/buildkit.(*Worker).setupRootfs(0x0?, {0x33d8038, 0x40014ee1e0}, 0x40038eac80)
        /app/engine/buildkit/executor_spec.go:416 +0x294
github.com/dagger/dagger/engine/buildkit.(*Worker).run(0x40069b00a0, {0x33d8038, 0x40014ee1e0}, 0x40038eac80, {0x4006623c28, 0x10, 0x29594a0?})
        /app/engine/buildkit/executor.go:222 +0x124
github.com/dagger/dagger/engine/buildkit.(*Worker).Run(_, {_, _}, {_, _}, {{_, _}, {_, _}, {_, ...}, ...}, ...)
        /app/engine/buildkit/executor.go:174 +0x40c
github.com/dagger/dagger/core.(*Service).startContainer.func7()
        /app/core/service.go:532 +0x16c
created by github.com/dagger/dagger/core.(*Service).startContainer in goroutine 105994
        /app/core/service.go:531 +0x2398
```

--- a/.changes/unreleased/Fixed-20251007-171917.yaml
+++ b/.changes/unreleased/Fixed-20251007-171917.yaml
@@ -1,0 +1,7 @@
+kind: Fixed
+body: Fixed engine crash when container sourced from private image was used with terminal
+  or as a service (and was not already cached in the engine).
+time: 2025-10-07T17:19:17.840364128-07:00
+custom:
+  Author: sipsma
+  PR: "11198"

--- a/core/integration/services_test.go
+++ b/core/integration/services_test.go
@@ -2287,7 +2287,7 @@ func (ServiceSuite) TestServiceFromUncachedPrivateImage(ctx context.Context, t *
 			"engine", "local-cache", "prune",
 		)).
 		With(daggerNonNestedExec("-c",
-			"container | from "+alpineImage+" | with-service-binding idc $(container | from "+privateImageRef+" | as-service) | with-exec apk add curl | with-exec curl http://idc | stdout",
+			"container | from "+alpineImage+" | with-service-binding idc $(container | from "+privateImageRef+" | with-exposed-port 80 | as-service) | with-exec apk add curl | with-exec curl http://idc | stdout",
 		))
 
 	out, err := ctr.Stdout(ctx)

--- a/core/integration/services_test.go
+++ b/core/integration/services_test.go
@@ -30,7 +30,9 @@ import (
 	"text/template"
 	"time"
 
+	bkconfig "github.com/dagger/dagger/internal/buildkit/cmd/buildkitd/config"
 	"github.com/dagger/dagger/internal/buildkit/identity"
+	resolverconfig "github.com/dagger/dagger/internal/buildkit/util/resolver/config"
 	"github.com/google/go-containerregistry/pkg/v1/tarball"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/sync/errgroup"
@@ -2229,6 +2231,68 @@ func (ServiceSuite) TestSearchDomainAlwaysSet(ctx context.Context, t *testctx.T)
 		}
 	}
 	require.True(t, found)
+}
+
+func (ServiceSuite) TestServiceFromUncachedPrivateImage(ctx context.Context, t *testctx.T) {
+	c := connect(ctx, t)
+
+	const htpasswd = "john:$2y$05$/iP8ud0Fs8o3NLlElyfVVOp6LesJl3oRLYoc3neArZKWX10OhynSC"
+	privateRegistrySvc := c.Container().
+		From("registry:2").
+		WithNewFile("/auth/htpasswd", htpasswd).
+		WithEnvVariable("REGISTRY_AUTH", "htpasswd").
+		WithEnvVariable("REGISTRY_AUTH_HTPASSWD_REALM", "Registry Realm").
+		WithEnvVariable("REGISTRY_AUTH_HTPASSWD_PATH", "/auth/htpasswd").
+		WithExposedPort(5000, dagger.ContainerWithExposedPortOpts{Protocol: dagger.NetworkProtocolTcp}).
+		AsService(dagger.ContainerAsServiceOpts{UseEntrypoint: true})
+
+	engineSvc := devEngineContainerAsService(devEngineContainer(c,
+		func(ctr *dagger.Container) *dagger.Container {
+			return ctr.
+				WithServiceBinding("registry", privateRegistrySvc)
+		},
+		engineWithBkConfig(ctx, t, func(ctx context.Context, t *testctx.T, cfg bkconfig.Config) bkconfig.Config {
+			cfg.Registries = map[string]resolverconfig.RegistryConfig{
+				"registry:5000": {
+					PlainHTTP: ptr(true),
+				},
+			}
+			return cfg
+		}),
+	))
+
+	const authFile = `{
+        "auths": {
+                "registry:5000": {
+                        "auth": "am9objp4RmxlamFQZGpydDI1RHZy"
+                }
+        }
+}`
+
+	privateImageRef := "registry:5000/test:latest"
+	ctr := engineClientContainer(ctx, t, c, engineSvc).
+		WithNewFile("/docker/config.json", authFile).
+		WithEnvVariable("DOCKER_CONFIG", "/docker").
+		With(daggerNonNestedExec("core",
+			"container",
+			"from", "--address", busyboxImage,
+			"with-workdir", "--path", "/srv",
+			"with-new-file", "--path", "index.html", "--contents", "yoyoyo",
+			"with-default-args", "--args", "httpd,-v,-f",
+			"with-exposed-port", "--port", "80",
+			"publish", "--address", privateImageRef,
+		)).
+		// prune the engine to ensure it's not cached no mo
+		With(daggerNonNestedExec("core",
+			"engine", "local-cache", "prune",
+		)).
+		With(daggerNonNestedExec("-c",
+			"container | from "+alpineImage+" | with-service-binding idc $(container | from "+privateImageRef+" | as-service) | with-exec apk add curl | with-exec curl http://idc | stdout",
+		))
+
+	out, err := ctr.Stdout(ctx)
+	require.NoError(t, err)
+	require.Equal(t, "yoyoyo", strings.TrimSpace(out))
 }
 
 func httpService(ctx context.Context, t testing.TB, c *dagger.Client, content string) (*dagger.Service, string) {

--- a/core/service.go
+++ b/core/service.go
@@ -453,9 +453,15 @@ func (svc *Service) startContainer(
 
 	name := fmt.Sprintf("container %s", svcID)
 	mm := bkmounts.NewMountManager(name, cache, session)
-	p, err := bkcontainer.PrepareMounts(ctx, mm, cache, bksession.NewGroup(bk.ID()), "", pbmounts, workerRefs, func(m *pb.Mount, ref bkcache.ImmutableRef) (bkcache.MutableRef, error) {
-		return cache.New(ctx, ref, nil)
+
+	bkSessionGroup := bksession.NewGroup(bk.ID())
+	p, err := bkcontainer.PrepareMounts(ctx, mm, cache, bkSessionGroup, "", pbmounts, workerRefs, func(m *pb.Mount, ref bkcache.ImmutableRef) (bkcache.MutableRef, error) {
+		return cache.New(ctx, ref, bkSessionGroup)
 	}, runtime.GOOS)
+	if err != nil {
+		return nil, fmt.Errorf("prepare mounts: %w", err)
+	}
+
 	for _, active := range slices.Backward(p.Actives) { // call in LIFO order
 		cleanup.Add("release active ref", func() error {
 			return active.Ref.Release(context.WithoutCancel(ctx))

